src/scvi/external/decipher/_module.py

```diff
@@ -0,0 +1,179 @@
+        )
+
+        with (
+            pyro.plate("batch", x.shape[0]),
+            poutine.scale(scale=1.0),
+        ):
+            with poutine.scale(scale=self.beta):
+                prior = dist.Normal(0, x.new_ones(self.dim_v)).to_event(1)
+                v = pyro.sample("v", prior)
+
+            z_loc, z_scale = self.decoder_v_to_z(v)
+            z_scale = softplus(z_scale)
+            z = pyro.sample("z", dist.Normal(z_loc, z_scale).to_event(1))
+
+            mu = self.decoder_z_to_x(z)
+            mu = softmax(mu, dim=-1)
+            library_size = x.sum(axis=-1, keepdim=True)
+            # Parametrization of Negative Binomial by the mean and inverse dispersion
+            # See https://github.com/pytorch/pytorch/issues/42449
+            # noinspection PyTypeChecker
+            logit = torch.log(library_size * mu + self._epsilon) - torch.log(
+                self.theta + self._epsilon
+            )
+            # noinspection PyUnresolvedReferences
+            x_dist = dist.NegativeBinomial(total_count=self.theta + self._epsilon, logits=logit)
+            pyro.sample("x", x_dist.to_event(1), obs=x)
+
+    @auto_move_data
+    def guide(self, x: torch.Tensor):
+        pyro.module("decipher", self)
+        with (
+            pyro.plate("batch", x.shape[0]),
+            poutine.scale(scale=1.0),
+        ):
+            x = torch.log1p(x)
+
+            z_loc, z_scale = self.encoder_x_to_z(x)
+            z_scale = softplus(z_scale)
+            posterior_z = dist.Normal(z_loc, z_scale).to_event(1)
+            z = pyro.sample("z", posterior_z)
+
+            zx = torch.cat([z, x], dim=-1)
+            v_loc, v_scale = self.encoder_zx_to_v(zx)
+            v_scale = softplus(v_scale)
+            with poutine.scale(scale=self.beta):
+                posterior_v = dist.Normal(v_loc, v_scale).to_event(1)
+                pyro.sample("v", posterior_v)
+        return z_loc, v_loc, z_scale, v_scale
+
+    def predictive_log_likelihood(self, x: torch.Tensor, n_samples=5):
+        """
+        Calculate the predictive log-likelihood for a Decipher module.
+
+        This function performs multiple runs through the dataloader to obtain
+        an empirical estimate of the predictive log-likelihood. It calculates the
+        log-likelihood for each run and returns the average. The beta parameter
+        of the Decipher module is temporarily modified and restored even if an
+        exception occurs. Used by default as an early stopping criterion.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Batch of data to compute the log-likelihood for.
+        n_samples : int, optional
+            Number of passes through the dataloader (default is 5).
+
+        Returns
+        -------
+        float
+            The average estimated predictive log-likelihood across multiple runs.
+        """
+        log_weights = []
+        old_beta = self.beta
+        self.beta = 1.0
```

Comment:
why not make beta an optional argument to model and guide and overwrite it then. This is cleaner.

Reply:
Why do you need a try block here?

Reply:
great idea, implemented this